p2p/net/network.go

```diff
@@ -303,20 +307,29 @@ func (n *Net) accept(listen net.Listener) {
 		conn := netConn.(*net.TCPConn)
 		n.tcpSocketConfig(conn) // TODO maybe only set this after session handshake to prevent denial of service with big messages
 		c := newConnection(netConn, n, nil, nil, n.config.MsgSizeLimit, n.config.ResponseTimeout, n.logger)
-		go func(con Connection) {
-			defer func() { pending <- struct{}{} }()
-			err := c.setupIncoming(n.config.SessionTimeout)
-			if err != nil {
-				n.logger.Event().Warning("conn_incoming_failed", log.String("remote", c.remoteAddr.String()), log.Err(err))
-				return
-			}
-			go c.beginEventProcessing()
-		}(c)
 
-		// network won't publish the connection before it the remote node had established a session
+		// Handle the incoming connection asynchronously
+		go n.acceptAsync(c, pending)
 	}
 }
 
+func (n *Net) acceptAsync(conn fmtConnection, pending chan struct{}) {
+	// Return the new connection token once the connection is closed
+	defer func() { pending <- struct{}{} }()
+
+	if err := conn.setupIncoming(n.config.SessionTimeout); err != nil {
+		n.logger.Event().Warning("conn_incoming_failed",
+			log.String("remote", conn.RemoteAddr().String()),
+			log.Err(err))
+		return
+	}
+
+	// The connection is automatically closed when there's no more to read, no need to close it here
+	conn.beginEventProcessing()
```

Comment:
isn't this a blocking function so the `defer` call above will never be executed until `beginEventProcessing` terminates which is basically when the connection is closed ?

Reply:
yes, it's blocking, that's why this whole function gets run inside a goroutine. Strictly speaking we don't need a defer here but I still think it's cleaner, more idiomatic go: this is a form of cleanup that should happen even if, e.g., there's a panic in `conn.beginEventProcessing()`. Let me know if you disagree.

Reply:
the defer works and idiomatic go because we have 2 return cases. though the actual problem I see is that we're not releasing `pending` channel until the connection is closed, this was not what this code was intended to do once the connection is running and open it is not pending anymore and we don't need to hold the channel.

Reply:
I agree about cleaning up and the panic case though.

Reply:
Not sure I understand. The reason this is a memory leak is that we were releasing `pending` before closing the connection. If we do this, there can be any number of such connections at one time. Look here, you can see there are hundreds of them: https://github.com/spacemeshos/go-spacemesh/issues/2313#issuecomment-808552000. This caused our managed miners to run out of memory and crash, as far as I can tell. That's why I modified this not to release until the connection closes. The current testnet is running with this code and seems stable!